aten/src/ATen/cuda/cub.cuh

```diff
@@ -51,6 +52,38 @@ struct cuda_type<c10::Half> {
   using type = __half;
 };
 
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 99999
+
+// waiting for https://github.com/NVIDIA/cub/pull/306 to land on CUDA
+template<>
+struct cuda_type<c10::BFloat16> {
+  using type = __nv_bfloat16;
+};
+
+#else
+
+template <>
+struct cub::FpLimits<c10::BFloat16>
+{
+    static __host__ __device__ __forceinline__ c10::BFloat16 Max() {
+        unsigned short max_word = 0x7F7F;
+        return reinterpret_cast<c10::BFloat16&>(max_word);
+    }
+
+    static __host__ __device__ __forceinline__ c10::BFloat16 Lowest() {
+        unsigned short lowest_word = 0xFF7F;
+        return reinterpret_cast<c10::BFloat16&>(lowest_word);
+    }
+};
+
+template <> struct cub::NumericTraits<c10::BFloat16>: cub::BaseTraits<cub::FLOATING_POINT, true, false, unsigned short, c10::BFloat16> {};
```

Comment:
@jeffdaily Is there any way to make this hack work on ROCm? Looks like in https://github.com/ROCmSoftwarePlatform/cub-hip/blob/hip_port_1.7.4/cub/util_type.cuh#L1102-L1194, there is no specialization for `__half`, so I think specializing for `c10::BFloat16` won't work either.

Reply:
OK, I was looking at an obsolete repo. The new repo does have `__half` specialization. But still, ROCm CI is failing:
```
May 13 02:15:01 In file included from /var/lib/jenkins/workspace/aten/src/ATen/native/hip/Randperm.hip:6:
May 13 02:15:01 /var/lib/jenkins/workspace/aten/src/ATen/hip/cub.cuh:89:18: error: explicit specialization of non-template struct 'FpLimits'
May 13 02:15:01 struct ::hipcub::FpLimits<c10::BFloat16>
May 13 02:15:01                  ^       ~~~~~~~~~~~~~~~
May 13 02:15:01 /var/lib/jenkins/workspace/aten/src/ATen/hip/cub.cuh:89:18: error: no struct named 'FpLimits' in namespace 'hipcub'
May 13 02:15:01 struct ::hipcub::FpLimits<c10::BFloat16>
May 13 02:15:01        ~~~~~~~~~~^
May 13 02:15:01 /var/lib/jenkins/workspace/aten/src/ATen/hip/cub.cuh:89:18: error: cannot define or redeclare 'FpLimits' here because namespace 'detail' does not enclose namespace 'hipcub'
May 13 02:15:01 struct ::hipcub::FpLimits<c10::BFloat16>
May 13 02:15:01        ~~~~~~~~~~^
May 13 02:15:01 /var/lib/jenkins/workspace/aten/src/ATen/hip/cub.cuh:102:30: error: explicit specialization of non-template struct 'NumericTraits'
May 13 02:15:01 template <> struct ::hipcub::NumericTraits<c10::BFloat16>: ::hipcub::BaseTraits<::hipcub::FLOATING_POINT, true, false, unsigned short, c10::BFloat16> {};
May 13 02:15:01                              ^            ~~~~~~~~~~~~~~~
May 13 02:15:01 /var/lib/jenkins/workspace/aten/src/ATen/hip/cub.cuh:102:30: error: no struct named 'NumericTraits' in namespace 'hipcub'
May 13 02:15:01 template <> struct ::hipcub::NumericTraits<c10::BFloat16>: ::hipcub::BaseTraits<::hipcub::FLOATING_POINT, true, false, unsigned short, c10::BFloat16> {};
May 13 02:15:01                    ~~~~~~~~~~^
May 13 02:15:01 /var/lib/jenkins/workspace/aten/src/ATen/hip/cub.cuh:102:30: error: cannot define or redeclare 'NumericTraits' here because namespace 'detail' does not enclose namespace 'hipcub'
May 13 02:15:01 template <> struct ::hipcub::NumericTraits<c10::BFloat16>: ::hipcub::BaseTraits<::hipcub::FLOATING_POINT, true, false, unsigned short, c10::BFloat16> {};
May 13 02:15:01                    ~~~~~~~~~~^
May 13 02:15:01 /var/lib/jenkins/workspace/aten/src/ATen/hip/cub.cuh:102:70: error: unknown template name 'BaseTraits'
May 13 02:15:01 template <> struct ::hipcub::NumericTraits<c10::BFloat16>: ::hipcub::BaseTraits<::hipcub::FLOATING_POINT, true, false, unsigned short, c10::BFloat16> {};
May 13 02:15:01                                                                      ^
May 13 02:15:01 /var/lib/jenkins/workspace/aten/src/ATen/hip/cub.cuh:102:91: error: no member named 'FLOATING_POINT' in namespace 'hipcub'
May 13 02:15:01 template <> struct ::hipcub::NumericTraits<c10::BFloat16>: ::hipcub::BaseTraits<::hipcub::FLOATING_POINT, true, false, unsigned short, c10::BFloat16> {};
May 13 02:15:01                                                                                 ~~~~~~~~~~^
May 13 02:15:01 8 errors generated when compiling for gfx900.
May 13 02:15:01 CMake Error at torch_hip_generated_Randperm.hip.o.cmake:192 (message):
May 13 02:15:01   Error generating file
May 13 02:15:01   /var/lib/jenkins/workspace/build/caffe2/CMakeFiles/torch_hip.dir/__/aten/src/ATen/native/hip/./torch_hip_generated_Randperm.hip.o
May 13 02:15:01 
```

Reply:
@zasdfgbnm ,
Can you please let us know how the compilation problem was fixed for ROCm build.

Reply:
It's not solved, bfloat16 is not compiled for rocm.

Reply:
Okay, that makes sense. It's okay to skip bfloat16 radix sort for ROCm. We'll look into adding bf16 support for radix sort into hipcub/rocprim.

Reply:
Internal issue to track adding bf16 support for radix sort into hipcub/rocprim: [SWDEV-288399](https://ontrack-internal.amd.com/browse/SWDEV-288399)